Update: Add fixer for prefer-const (fixes #6448)

NickHeiner opened the pull request:
This time, I remembered to update the docs 😄 !

@nzakas told me that it's ok if the fixer does not catch every case. I tried to stay away from the cases where it would be less obvious how the user would want to have the issue be fixed; let me know if you think are some more cases I should try to handle.

I also improved the grammar on the error message for `prefer-const`.

ilyavolodin commented:
Hmm... This is interesting. This is the first rule (I think) that we have that doesn't fix all use cases. How do we do negative test cases?

mysticatea commented on lib/rules/prefer-const.js:
```diff
@@ -240,11 +259,32 @@ module.exports = {
+                        node.parent.parent.type === "ForStatement" ||
+                        node.parent.init),
+                isDestructuring = node.parent.parent.type === "ObjectPattern" &&
+                    node.parent.parent.parent.parent.type === "VariableDeclaration",
```
I'm not sure the intention of this `isDestructuring`.
What is this condition?

NickHeiner replied:
Comment added to explain.

mysticatea replied:
Thank you, but I'm still not sure.
Why do you fix `let {a, b, c}`, but not `let [a, b, c]` or `let {a: {b, c}}` ?
Also, the variable name `isDestructuring` is not expressing this condition.

mysticatea commented on lib/rules/prefer-const.js:
```diff
@@ -240,11 +259,32 @@ module.exports = {
+                isNormalVariableDeclaration = node.parent.parent.type === "VariableDeclaration" &&
+                    (node.parent.parent.parent.type === "ForInStatement" ||
+                        node.parent.parent.parent.type === "ForOfStatement" ||
+                        node.parent.parent.type === "ForStatement" ||
```
`node.parent.parent.type === "VariableDeclaration" && ... node.parent.parent.type === "ForStatement" ...`?

NickHeiner replied:
I simplified this logic.

mysticatea commented on lib/rules/prefer-const.js:
```diff
@@ -240,11 +259,32 @@ module.exports = {
+
+            if (isNormalVariableDeclaration || isDestructuring) {
+
+                reportArgs.fix = function(fixer) {
```
I'm not sure this is proper.

Because this `report()` function is called for each identifier node, and so this function may be called multiple times for a variable declaration node. I guess it needs to check whether all identifiers of this declaration are not reassigned.

NickHeiner replied:
Are you talking about a case like this?

```js
let {a, b} = obj; // b should be const
a = 1;
```

I was going to leave this case unfixed, because it's not obvious how a developer would want to resolve the warning.

mysticatea replied:
Yes.
Actually, your example code is fixed to `const` currently.

mysticatea replied:
Yes.
Actually, your example code is fixed to `const` currently.

mysticatea replied:
Also, this is similar.

```js
let a = 0, b = 0;
a = 1;
```

NickHeiner commented:
> How do we do negative test cases?

Well, we have test cases where there is no `output` in the object returned by the rule. For example:

```js
        {
            code: "let [x = -1, y] = [1,2]; y = 0;",
            parserOptions: { ecmaVersion: 6 },
            errors: [{ message: "'x' is never reassigned, use 'const' instead.", type: "Identifier"}]
        },
```

Is that what you're looking for?

ilyavolodin commented:
@NickHeiner Problem is, that if no `output` is provided, we don't check `fix` at all. I don't think we currently have a way to verify that `fix` will NOT be applied to a given test case. That's what I mean by negative tests.

platinumazure commented:
@ilyavolodin Can he just copy the text code into output? That might work short-term. Long-term, we can explore better options.

NickHeiner commented:
Ah, I see what you're saying @ilyavolodin. I'll take a look.

NickHeiner commented:
So we could update `rule-tester` to have the following:

```diff
diff --git a/lib/testers/rule-tester.js b/lib/testers/rule-tester.js
index 2ee87ec..6ded833 100644
--- a/lib/testers/rule-tester.js
+++ b/lib/testers/rule-tester.js
@@ -429,6 +429,14 @@ RuleTester.prototype = {
                     var fixResult = SourceCodeFixer.applyFixes(eslint.getSourceCode(), messages);
 
                     assert.equal(fixResult.output, item.output, "Output is incorrect.");
+                } else {
+                    var noFixResult = SourceCodeFixer.applyFixes(eslint.getSourceCode(), messages);
+
+                    assert.equal(
+                        noFixResult.output,
+                        item.code,
+                        "If the test item does not have an output property, then applying fixes should be a no-op."
+                    );
                 }
 
             }

```

This causes some other tests to fail, but I like the idea that a test must be explicit if a fixer exists about what the fixer does. What do you think? If you like it, I'll go through and make the other tests pass.

mysticatea commented:
I agree with https://github.com/eslint/eslint/pull/6486#issuecomment-227508861
If some tests have `output` same as `code`, it's meaning that the code of those tests are not modified. It's explicit.

mysticatea commented on lib/rules/prefer-const.js:
```diff
@@ -240,11 +259,33 @@ module.exports = {
+                },
+                isNormalVariableDeclaration = node.parent.parent.parent.type === "ForInStatement" ||
+                        node.parent.parent.parent.type === "ForOfStatement" ||
+                        node.parent.parent.type === "ForStatement" ||
```
What is your intention of `node.parent.parent.type === "ForStatement"`?

NickHeiner replied:
It was there erroneously! Good catch.

mysticatea commented on lib/rules/prefer-const.js:
```diff
@@ -240,11 +259,33 @@ module.exports = {
+                // {let {a, b} = obj} should be written as {const {a, b} = obj}
+                isDestructuring = node.parent.parent.type === "ObjectPattern",
+
+                variableDeclarationParent = findUp(node, "VariableDeclaration");
```
This would cause unintentional behavior.
e.g.

```js
var foo = function() {  // fixed `var foo` to `const foo`
    let a;
    ({a} = 1);
}
```

```js
var foo = function() {  // fixed `var foo` to `const foo`
    for (const b of c) { 
        let a; 
        a = 1; 
    } 
}
```

I think `findUp` should stop at `*Statement` node.
Also, the found `VariableDeclaration` node should be used for other conditions instead of `node.parent.parent`.

mysticatea replied:
Strange.
I put the code after I confirmed it works.

NickHeiner replied:
Ah yep, I was using Node v4. I see that it works in Node v6.

NickHeiner commented:
> If some tests have output same as code, it's meaning that the code of those tests are not modified. It's explicit.

Ok, I'll just take that approach then.

NickHeiner commented:
@mysticatea out of curiosity, why does this rule only warn on `let` but not `var`?

platinumazure commented:
@NickHeiner My guess is due to `var` hoisting.

Take this example:

```js
function myFunc(foo) {
    if (foo) {
        var bar = true;
    }
    console.log(bar);
}
```

Switching `var` to `const` in that example will break the code. (Of course, one could reasonably argue that the developer should know how `var` works and should hoist it themselves, but the point is a linter should not auto-fix in such a way that it introduces more errors, if possible.)

nzakas commented:
Yes, we only warn on `let` because `const` has the same scoping rules while `var` has different scoping rules.

nzakas commented on tests/lib/rules/prefer-const.js:
```diff
@@ -94,140 +94,190 @@ ruleTester.run("prefer-const", rule, {
             code: "(function() { let {a: x = -1, b: y} = {a:1,b:2}; y = 0; })();",
             parserOptions: { ecmaVersion: 6 },
-            errors: [{ message: "'x' is never reassigned, use 'const' instead.", type: "Identifier"}]
+            errors: [{ message: "'x' is never reassigned. Use 'const' instead.", type: "Identifier"}]
```
Missing `output` for this case

nzakas commented:
Overall looks good. @NickHeiner a bit of feedback: changing the message in such a minor way is a bit annoying for reviewers. It makes the diff a lot noisier for no good reason. Best to not make small tweaks like that when they're not directly necessary for the task.

NickHeiner commented:
@platinumazure @nzakas that makes sense about the `var` hoisting.

@nzakas: I'm assuming that you're referring to me changing the warning message. I know that what I did substantially widens the diff, and if you want me to remove that change, I will. For what it's worth, I didn't make the change lightly; I considered it to be bringing grammatical correctness (avoiding a run-on) instead of just a stylistic modification. But I also know that I'm more pedantic than most when it comes to things like this, so I'm happy to drop it, and I won't suggest a change like this in the future unless I see the error as egregious :smile:.

platinumazure commented on lib/rules/prefer-const.js:
```diff
@@ -203,6 +203,23 @@ function groupByDestructuring(variables, ignoreReadBeforeAssign) {
+    if (node.type === type) {
+        return node;
+    }
+    return findUp(node.parent, type);
```
Is this worth turning into an iterative form? Normally I don't raise a huge stink about recursion, but this one seems pretty easy to turn into a `while` loop.

NickHeiner replied:
I tend to prefer recursion as a matter of style, but I'm happy to do this iteratively. If you feel strongly, I'll change it.

platinumazure commented:
@nzakas I occasionally run into sentence which I have to force myself not to correct, so I guess I'd like to know: Are you saying you don't want text changes in PRs that focus primarily on functionality changes (because they add to the diff and confuse things) but that you're okay with text changes proposed as separate issues/PRs focuses exclusively on the text? Or that you would rather we do not propose text changes at all without a very good reason?

mysticatea commented:
> out of curiosity, why does this rule only warn on let but not var?

Yeah, `var` is warned by `no-var` rule.

--

I have some concerns.

> Thank you, but I'm still not sure.
> Why do you fix `let {a, b, c}`, but not `let [a, b, c]` or `let {a: {b, c}}` ?
> Also, the variable name `isDestructuring` is not expressing this condition.
> https://github.com/eslint/eslint/pull/6486#discussion_r68031882

-

> Yes.
> Actually, your example code is fixed to const currently.
> https://github.com/eslint/eslint/pull/6486#discussion_r68028631

-

> I think `findUp` should stop at `*Statement` node.
> https://github.com/eslint/eslint/pull/6486#discussion_r68034310

NickHeiner commented:
Sorry for missing your comments @mysticatea. I've pushed up a few more test cases based on what you wrote, but I have a few more fixes to investigate tomorrow morning.

NickHeiner commented:
Thanks for you feedback @mysticatea. I've pushed up additional test cases and modifications based on what you wrote. Hopefully now I've addressed all comments.

mysticatea commented:
@NickHeiner Thank you for your great work! Looks good to me now.

A nit picking, variable name `_` is not preferred in this project. Please use `lodash` instead.

NickHeiner commented:
Thanks for the review @mysticatea. I tried to rebase and also pull in the latest changes from master, but I screwed it up. Please do not merge as is -- I'll get this fixed soon.

NickHeiner commented:
Ok, branch errors fixed. This should be good to go, pending any other comments.

ilyavolodin commented:
LGTM

nzakas commented:
Lgtm. Thanks @NickHeiner !
